pkg/volumeclaimbinder/persistent_volume_claim_binder.go

```diff
@@ -0,0 +1,301 @@
+			volume.Status.Phase = api.VolumeBound
+			_, err := binderClient.UpdatePersistentVolumeStatus(volume)
+			if err != nil {
+				return fmt.Errorf("Error updating pv.status: %v\n", err)
+			}
+		}
+
+		// verify the volume is still claimed by a user
+		if claim, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name); err == nil {
+			glog.V(5).Infof("PersistentVolume[%s] is bound to PersistentVolumeClaim[%s]\n", volume.Name, volume.Spec.ClaimRef.Name)
+			// rebuild the Claim's Status as needed
+			if claim.Status.VolumeRef == nil {
+				syncClaimStatus(binderClient, volume, claim)
+			}
+		} else {
+			//claim was deleted by user.
+			glog.V(3).Infof("PersistentVolumeClaim[%s] unbound from PersistentVolume[%s]\n", volume.Spec.ClaimRef.Name, volume.Name)
+			// volume.Spec.ClaimRef is deliberately left non-nil so that another process can recycle the newly released volume
+			volume.Status.Phase = api.VolumeReleased
+			volume, err = binderClient.UpdatePersistentVolumeStatus(volume)
+			if err != nil {
+				return fmt.Errorf("Error updating pv: %+v\n", err)
+			}
+		}
+	} else {
+		volume.Status.Phase = api.VolumeAvailable
+		_, err := binderClient.UpdatePersistentVolumeStatus(volume)
+		if err != nil {
+			return fmt.Errorf("Error updating pv.status: %v\n", err)
+		}
+	}
+	return nil
+}
+
+func syncClaimStatus(binderClient binderClient, volume *api.PersistentVolume, claim *api.PersistentVolumeClaim) (err error) {
+	volumeRef, err := api.GetReference(volume)
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting volume reference: %v\n", err)
+	}
+	// all "actuals" are transferred from PV to PVC so the user knows what
+	// type of volume they actually got for their claim
+	claim.Status.Phase = api.ClaimBound
+	claim.Status.VolumeRef = volumeRef
+	claim.Status.AccessModes = volume.Spec.AccessModes
+	claim.Status.Capacity = volume.Spec.Capacity
+
+	_, err = binderClient.UpdatePersistentVolumeClaimStatus(claim)
+
+	if err != nil {
+		claim.Status.Phase = api.ClaimPending
+		claim.Status.VolumeRef = nil
+		claim.Status.AccessModes = nil
+		claim.Status.Capacity = nil
+	}
+
+	return err
+}
+
+func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderClient, claim *api.PersistentVolumeClaim) (err error) {
+	glog.V(5).Infof("Synchronizing PersistentVolumeClaim[%s]\n", claim.Name)
+
+	if claim.Status.VolumeRef != nil {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] is bound to PersistentVolume[%s]\n", claim.Name, claim.Status.VolumeRef.Name)
+		return nil
+	}
+
+	volume, err := volumeIndex.FindBestMatchForClaim(claim)
+	if err != nil {
+		return err
+	}
+
+	if volume != nil {
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
+		// make a binding reference to the claim
+		volume.Spec.ClaimRef = claimRef
+		volume, err = binderClient.UpdatePersistentVolume(volume)
```

Comment:
Covering this whole block - I am not sure in reading this that rollback of this multi-object transaction is right.  I know it's an annoying corner case, but I think it needs to be clearer what is happening and that this is on purpose.

Really good comments would help.

Being more explicit about what gets reverted would help.

Something like a defer'ed rollback or something would make it clearer.  Maybe something like:

https://github.com/GoogleCloudPlatform/kubernetes/blob/cd6daae014f8d3715f7305b49139df7210f1902a/pkg/registry/service/rest.go#L97-L121